sumologic/resource_sumologic_content.go

```diff
@@ -39,6 +41,73 @@ func resourceSumologicContent() *schema.Resource {
 	}
 }
 
+func configStateFunc(value interface{}) string {
+	return normalizeConfig(value.(string))
+}
+
+// modify json config to remove logically equivalent changes in terrafrom diff output
+// e.g. absent map entry vs map entry with null value
+// or absent map entry vs map entry with default value
+func normalizeConfig(originalConfig string) string {
+	config, err := structure.ExpandJsonFromString(originalConfig)
+
+	if err != nil {
+		log.Println("Couldn't expand config json from string")
+		return originalConfig
+	}
+
+	removeEmptyValues(config)
+	fillPanelQueriesDefaultValues(config)
+
+	if config["theme"] != nil {
+		config["theme"] = strings.ToLower(config["theme"].(string))
+	}
+
+	children, ok := config["children"].([]interface{})
+	if ok {
+		for _, childItemObject := range children {
+			childItemMap, ok := childItemObject.(map[string]interface{})
+			if ok {
+				fillPanelQueriesDefaultValues(childItemMap)
+			}
+		}
+	}
+	configString, err := structure.FlattenJsonToString(config)
+	if err != nil {
+		log.Println("Couldn't flatten config json to string")
+		return originalConfig
+	}
+
+	return configString
+}
+
+func fillPanelQueriesDefaultValues(config map[string]interface{}) {
+	if config["panels"] != nil {
+		panels := config["panels"].([]interface{})
+
+		for _, panelInterface := range panels {
+			panelItem := panelInterface.(map[string]interface{})
+
+			for range panelItem {
+				if panelItem["queries"] != nil {
+					queries := panelItem["queries"].([]interface{})
+					for _, queryInterface := range queries {
+						queryItem := queryInterface.(map[string]interface{})
+
+						if queryItem["outputCardinalityLimit"] == nil {
+							queryItem["outputCardinalityLimit"] = 1000
+						}
+
+						if queryItem["transient"] == nil {
+							queryItem["transient"] = false
+						}
```

Comment:
How are these used?

Reply:
Values of these fields (e.g. `outputCardinalityLimit`) are often unspecified in terraform. JSON returned from the Sumologic API contains these fields with default values (e.g. `"outputCardinalityLimit": 1000`). As a result we get a lot of diff outputs like this:
```
                         ~ {
                                id                                     = null
                              ~ queries                                = [
                                  ~ {
                                      - outputCardinalityLimit = 1000
                                      - spansQueryData         = null
                                      - transient              = false
                                        # (8 unchanged attributes hidden)
                                    },
                                ]
                                # (9 unchanged attributes hidden)
```
My aim here was to get rid of this diff that doesn't bring any value.